File: 30 days of code/Day 12 - Inheritance/Inheritance.cs
/**
 *              INHERITANCE | C#
 * Domain       30 DAYS OF CODE
 * Sub domain   DAY 12
 */

using System;
using System.IO;

class Program
{
    //static TextReader read = Console.In;
    static StreamReader read = new StreamReader("input.txt");

    static void Main(string[] args)
    {
        string[] line = read.ReadLine().Split(' ');
        string firstName = line[0];
        string lastName = line[1];
        int id = int.Parse(line[2]);
        int scores = int.Parse(read.ReadLine());
        int[] testScores = new int[scores];
        line = read.ReadLine().Split(' ');
        testScores = Array.ConvertAll(line, int.Parse);
        read.Close();

        Student student = new Student(firstName, lastName, id, testScores);
        student.printPerson();
        Console.WriteLine("Grade: " + student.Calculate());
    }
}

class Person
{
    private string FirstName { get; set; }
    private string LastName { get; set; }
    private int Id { get; set; }

    public Person() { }

    public Person(string firstName, string lastName, int id)
    {
        FirstName = firstName;
        LastName = lastName;
        Id = id;
    }

    public void printPerson()
    {
        Console.WriteLine("Name: " + LastName + ", " + FirstName + "\nID: " + Id);
    }
}

class Student : Person
{
    private int[] TestScores { get; set; }

    public Student(String firstName, String lastName, int id, int[] testScores) 
        : base(firstName, lastName, id)
    {
        TestScores = testScores;
    }

    public char Calculate()
    {
        int total = TestScores.Length;
        foreach (int i in TestScores) {
            total += i;
        }

        int mean = total / TestScores.Length;

        if (mean >= 90 && mean <= 100)
        {
            return 'O';
        }
        else if (mean >= 80)
        {
            return 'E';
        }
        else if (mean >= 70)
        {
            return 'A';
        }
        else if (mean >= 55)
        {
            return 'P';
        }
        else if (mean >= 40)
        {
            return 'D';
        }
        else
        {
            return 'T';
        }
    }
}

File: 30 days of code/Day 12 - Inheritance/Inheritance.cpp
/**
*              INHERITANCE | C++
* Domain       30 DAYS OF CODE
* Sub domain   DAY 12
*/

#include "stdafx.h"
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

using namespace std;

ifstream file("input.txt");
istream& f = file;

class Person {
public:
	Person(string firstName, string lastName, int id) {
		this->firstName = firstName;
		this->lastName = lastName;
		this->id = id;
	}

	void printPerson() {
		cout << "Name" << lastName << ", " << firstName << "\nID: " << id << "\n";
	}

private:
	string firstName;
	string lastName;
	int id;
};

class Student : public Person {
public:
	Student(string firstName, string lastName, int id, vector<int> testScores)
		: Person(firstName, lastName, id) {
		this->testScores = testScores;
	}

	char calculate() {
		int total = 0;
		for (int i : testScores) {
			total += i;
		}

		int mean = total / testScores.size();

		if (mean >= 90 && mean <= 100) {
			return 'O';
		} else if (mean >= 80) {
			return 'E';
		} else if (mean >= 70) {
			return 'A';
		} else if (mean >= 55) {
			return 'P';
		} else if (mean >= 40) {
			return 'D';
		} else {
			return 'T';
		}
	}

private:
	vector<int> testScores;
};

void main() {
	string firstName; f >> firstName;
	string lastName; f >> lastName;
	int id; f >> id;
	int scores; f >> scores;
	vector<int> testScores;
	for (int i = 0; i < scores; ++i) {
		int temp; f >> temp;
		testScores.push_back(temp);
	}

	Student* student = new Student(firstName, lastName, id, testScores);
	student->printPerson();
	cout << "Grade: " << student->calculate() << endl;
}